test/customer.js

```diff
@@ -64,145 +64,270 @@ describe('Customer API:', () => {
     expect(requests.length).to.equal(0);
   });
 
-  describe('when customerId is unknown and externalId is not provided,', () => {
-    beforeEach(() => {
-      setConfig();
-      _ch('customer', mario);
+  describe('when a customerId is not provided', () => {
+    describe('and no customerId is stored in the cookie', () => {
+      beforeEach(() => {
+        setConfig();
+        _ch('customer', mario);
+      });
+
+      it('creates a new customer', (done) => {
+        whenDone(() => {
+          expect(requests.length).to.equal(1);
+          const req = requests[0];
+          expect(req.method).to.equal('POST');
+          expect(req.url).to.equal(
+            `${apiUrl}/workspaces/${config.workspaceId}/customers`
+          );
+          expect(JSON.parse(req.requestBody)).to.eql({
+            nodeId: config.nodeId,
+            externalId: mario.externalId,
+            base: mario.base
+          });
+          expect(req.requestHeaders.Authorization).to.equal(
+            `Bearer ${config.token}`
+          );
+          done();
+        });
+      });
+
+      it('handles 409 conflicts and updates the existing customer', (done) => {
+        whenDone(() => {
+          requests[0].respond(409, {}, JSON.stringify({
+            data: {
+              customer: {
+                id: 'existing-cid'
+              }
+            }
+          }));
+          whenDone(() => {
+            expect(requests.length).to.equal(2);
+            const req = requests[1];
+            expect(req.method).to.equal('PATCH');
+            expect(req.url).to.equal(
+              `${apiUrl}/workspaces/${config.workspaceId}/customers/existing-cid`
+            );
+            expect(JSON.parse(req.requestBody)).to.eql({
+              externalId: mario.externalId,
+              base: mario.base
+            });
+            done();
+          });
+        });
+      });
+
+      it('stores the customerId for future calls', (done) => {
+        whenDone(() => {
+          requests[0].respond(200, {}, JSON.stringify({ id: 'new-cid' }));
+          whenDone(() => {
+            expect(getCookie().customerId).to.equal('new-cid');
+            done();
+          });
+        });
+      });
+
+      it('stores a hash of the customer data for future calls', (done) => {
+        whenDone(() => {
+          requests[0].respond(200, {}, JSON.stringify({ id: 'new-cid' }));
+          whenDone(() => {
+            expect(getCookie().hash).not.to.be.undefined;
```

Comment:
These lines were not changed in this PR but only moved. I'll leave them as they are for now to minimise the changes.